### PR TITLE
create interpreter implementation and adding instruction decoder for riscvi32

### PR DESCRIPTION
**Pull request by svv232**

### CODE REVIEW

File: o1vm/src/interpreters/riscv32i/witness.rs [hunk 1]
```diff
@@ -567,6 +567,64 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
...
+                general_purpose: state.registers,
+                current_instruction_pointer: initial_instruction_pointer,
+                next_instruction_pointer,
+                heap_pointer: state.heap,
+            }
+        };
+
+        let mut registers = initial_registers.clone();
+        registers[2] = 0x408004f0;
+        // set the stack pointer to the top of the stack
```
Comment (svv232) [inline o1vm/src/interpreters/riscv32i/witness.rs:604]: a follow up pr needs to be made to dynamically set the stack ptr based on the loader, this should be fine for now in the vms memory implementation.

Comment (dannywillems) [reply o1vm/src/interpreters/riscv32i/witness.rs:607]: Linking comment https://github.com/o1-labs/proof-systems/pull/2741

File: o1vm/src/interpreters/riscv32i/witness.rs [hunk 2]
```diff
@@ -567,6 +570,201 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
...
+        /* https://www.cs.cornell.edu/courses/cs3410/2024fa/assignments/cpusim/riscv-instructions.pdf */
+        let instruction =
+            ((self.get_memory_direct(self.registers.current_instruction_pointer) as u32) << 24)
+                | ((self.get_memory_direct(self.registers.current_instruction_pointer + 1) as u32)
+                    << 16)
+                | ((self.get_memory_direct(self.registers.current_instruction_pointer + 2) as u32)
+                    << 8)
+                | (self.get_memory_direct(self.registers.current_instruction_pointer + 3) as u32);
+        let instruction = instruction.to_be(); // convert to big endian for more straightforward decoding
+        println!(
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32i/witness.rs:641]: Can we use a debug! instead?

**Review by dannywillems — APPROVED** [review]
LGTM

File: o1vm/src/interpreters/riscv32i/witness.rs [hunk 3]
```diff
@@ -567,6 +570,201 @@ impl<Fp: Field> InterpreterEnv for Env<Fp> {
...
+    }
+
+    pub fn next_instruction_counter(&self) -> u64 {
+        (self.normalized_instruction_counter() + 1) * MAX_ACC
+    }
+
+    pub fn decode_instruction(&mut self) -> (Instruction, u32) {
+        /* https://www.cs.cornell.edu/courses/cs3410/2024fa/assignments/cpusim/riscv-instructions.pdf */
+        let instruction =
+            ((self.get_memory_direct(self.registers.current_instruction_pointer) as u32) << 24)
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32i/witness.rs:634]: Could we not avoid `to_be` below and inverse here the order of the bitshift? like first getting 0, 8, 16 and 24?

**Pull request merged**